README.md
# ctadmtool Package

This project is dedicated to perform analysis with `ctools` and `gammalib` to compute upper limits of annihilation cross-section and decay lifetime of dark matter particles using CTA observations.

The code presented here is functional at least for versions of `ctools` and `gammalib` greater than 1.6.3.

To install `ctools` and `gammalib` you can use conda. Also, you can build from source. Please check [gammalib](http://cta.irap.omp.eu/gammalib/admin/index.html) and [ctools](http://cta.irap.omp.eu/ctools/admin/index.html) pages.

The core of the analysis tool is `csdmatter`. This class is based in `csscripts` installed with ctools. One of the key features is the management of dark-matter spectrum via the `gammalib` `GModelSpectralTable` class. This has the advantage to only use one fits file to describe the spectrum for candidates with masses in a determined range, dedicated annihilation channels (following the convention from the [PPPC4DMID](http://www.marcocirelli.net/PPPC4DMID.html) project) and particular energy ranges. You can check the jupyter notebook `tablemodel` under the **notebooks/** folder to check how to create a spectrum fits file. The `tablemodel` notebook uses `dmtable` class to interpolate the spectrum for different masses, energy and channels using `dmspectrum` class. The spectrum is save into a fits table that is esaily ingested by `ctools` and `gammalib`.

The fits table has three parameters to describe the spectrum:

1. Mass. The mass of the dark matter candidate (in GeV). [**Default: fixed**]. In the case of a hypothetical dark-matter signal, you should free this parameter.
2. Channel. Annihilation channel. I took the same number convetion as in the PPPC4DMID project. [**Default: fixed**]. This variable should not set free during the analysis.
3. Normalization. Overall normalization of the dark-matter spectra. [**Default: free**]

Aditionally, you can select whether or not to include electroweak (EW) corrections in the spectrum. By default, this parameter is set to `True`.

## Current known issues:

1. Compute gamma-ray spectrum of decay to W's of WIMPs with masses below 160 GeV

If you found any issue during test or use of this project, please open an Issue.

## Description

The contents of the project are:

2.  **ctadmtool**
  * *data*
  * *dmspectrum*
  * *pfiles*
  * *tools*
  * **csdmatter.py**
4.  **LICENSE**
5.  **MANIFEST.in**
6.  **README.md**
7.  **setup.cfg**
8.  **setup.py**

## ctools and gammalib setup

You can set the ```GAMMALIB``` and ```CTOOLS``` environment variables as usual. See the documentation about gammalib and ctools.

##  The ctadmtool python package

`ctadmtool` is a python package to compute exclusion limits for model-independent dark matter searches with CTA. The package is an effort to have a common set of tools and use as basic example for analysis. There are two subpackages:

1. **dmspectrum**
2. **tools**

The relevan classes (about physics) are:

1.  *dmspectra*
  - To compute the number of photons produced during annihilation or decay of dark matter particles using PPPC4DMID tables. Here, it is also included EBL attenuation using [ebl-table project](https://github.com/me-manu/ebltable)
2.  *dmflux_table*
  - To generate `GModelSpectralTable` models for annihilation or decay of dark matter particles.

There are also some files in *data* and *pfiles* folders:

1.  *data*
  - Tables from PPPC4DMID project
2.  *pfiles*
  - Parameter file of **csdmatter** app
  - Help of **csdmatter** app

Finally, **ctadmtool** contains also the **csdmatter** app, based on cscripts.

##  The csdmatter app

The csdmatter app is based on how the cscripts are implemented within ctools. The csdmatter computes the upper-limits (at this moment, just) for annihilation cross-section for a famlily of mass points of dark matter particles. You can refer to [*pfiles/csdmatter.par*](ctadmtool/pfiles/csdmatter.par) and [*pfiles/csdmatter.txt*](ctadmtool/pfiles/csdmatter.txt) to check the full list of input parameters, and the help of the app.

##  Installation

To have **ctadmtool** package availabe in your system you must to be sure that *ctools* and *gammalib* are loaded. Then to install **ctadmtool** you have two options:

1. Cloning:
  - `$ git clone git@github.com:sergiohcdna/ctadmtool.git`
  - `$ cd ctadmtool`
  - `$ python -m pip install .`

2. Using pip directly:
  - `$ python -m pip install git@github.com:sergiohcdna/ctadmtool.git`

Please note, that, if you want to contribute to the development of **csdmatter** and related classes, you must use the first option. Additionally, you can create a branch.

**Note:** There is `Deprecation` message when usin `python -m pip install .` referring that the local packages will be building in-place. You can take a look at the [discussion](https://github.com/pypa/pip/issues/7555). To avoid the deprecation message and test this new feature you must use:

```
$ python -m pip install . --use-feature=in-tree-build
```

**Note:** If you are updating a previous installation of `ctadmtool`, please be sure that you don't have any old `csdmatter.par` files in other locations. This will create a problem if new parameters were added to the version you are trying to install. A manual solution is to erase the `csdmatter.par` files from the `$CTOOLS/syspfiles` and your `$HOME`:

```bash
$ rm $CTOOLS/syspfiles/csdmatter.par
$ rm $HOME/pfiles/csdmatter.par
```

After this, you can install `ctadmtool`.

### Running the csdmatter app

Because the script is not part of the default cscripts, you must execute it using inside a python script.

### DM Limits Calculation

To compute the ULs, **csdmatter** compute the ratio between the expected integrated flux and the upper-limit integrated flux obtained during the fit. Both fluxes are computed between minimum energy `emin` and 95% of the available energy in the process. The ratio is used to compute the exclusion limit using the reference annihilation cross-section or the decay lifetime used to compute the normalization of the dark matter spectrum. The ULs are computed for masses points (*mnumpoints* parameter) in the range of masses indicated by the user (via *mmin* and *mmax* parameters)

### Results

As already it was mentioned, the **csdmatter** app computes the upper-limits for a family of mass points (you can specify as many points as you need using the input parameter *mnumpoints*). These masses corresponds to differentes dark matter particles. They are separated logarithmically. For every mass value, the **csdmatter** app generate the corresponding GModel, and compute the upper-limit. Then, for every mass point, the following results are saved:

1. MinEnergy &#10140; Minimum Energy
2. MaxEnergy &#10140; Maximum Energy
3. Mass &#10140; Mass of the dark matter candidate
4. Flux &#10140; Flux at reference energy
5. ErrFlux &#10140; Error associated to flux
6. E2Flux &#10140; Energy squared times flux
7. E2ErrFlux &#10140; Energy squared times flux error
6. LogL &#10140; Log-Likelihood obtained during the fit
7. TS &#10140; Test Statistic
8. UpperLimit &#10140; Upper limit to the flux
11. ScaleFactor &#10140; Ratio between theoretical and upper-limit flux
12. ULLifetime or ULCrossSection &#10140; Exclusion Limit to Lifetime or Cross-section
13. RefLifetime or RefCrossSection &#10140; Reference values used to compute the dark matter flux

At the end, the results for all mass points are saved into a fits file.

If you use the `run` method, the results table can be accessed via the `dmatter_fits` method. For example:

```python
from ctadmtool.csdmatter import csdmatter

thistool = csdmatter()
# After all the parameter initialization

thistools.run()
results = thistool.dmatter_fits()
print(results)
```

You can take a look at the jupyter notebooks `get_decayllimits` and `get_dmulimits` to learn how to run the `csdmatter` tool to get exclusion limits for a *Toy dark-matter halo*.

## Notebooks

There are several jupyter notebooks to show you how to use the package. The notebooks are:

- `dmspectrum` &#10140; Shows how to use the`dmspectrum` class to get dark matter spectrum for annihilation or decay
- `tablemodel` &#10140; Shows how to use `dmtable` class to create `GModelSpectralTable` models for dark matter
- `plottingdmSpectra`&#10140; Shows how to plot spectra from table models
- `dmflux` &#10140; Shows how to compute gamma-ray fluxes using `dmspectrum` class. comparison with *Clumpy* gamma-ray flux for decay.

CHANGELOG
# Change Log
All notable changes to this project will be documented in this file.
 
The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

[Sergio]

## [0.1.1] - 2021-06-08
  
Long explanation of commits added in ctaAnalysis v0.1.1
 
### Added

- examples/create_dmtable.py  : Show how-to use GModelSpectralTable for dark matter spectral models
- notebooks/dmspectraltable   : Light version of examples/create_dmtable.py
- notebooks/loaddmspectra     : Load PPPC4DMID tables and save to a fits file
- notebooks/plottingdmSpectra : Plot spectra. Use fits file from loaddmspectra notebook

### Changed
  
- [gammalib models]      : Avoid use of XML constructors when creating spectral models
- [dm spectral model]    : Use gammalib.GModelSpectralTable to create multiple dm-spectral models on the fly
- [csdmatter parameters] : Add dmspecfits option to pass a fits file obtained from create_dmtable script

### Fixed
 
- [multiprocessing] : Using GVector as container for masses points causes the following error "python TypeError: can't pickle SwigPyObject objects". Change masses container to a list.


[Sergio]
#   Compute DM decay limits
## List of changes from update 1012,2021

1. dmspectrum class
    1. add decay as option to compute dm spectra
    2. calculation of decay spectrum
    3. add emin, emax and epoints as parameters to compute energy range where the spectrum is computed
    4. Add properties emin and emax. Setters update the list used to compute the spectrum
    5. aAdd energy setter with tuple with emin, emax, epoints as input parameter
2. notebooks/
    1. dmspectrum —> Show how-to use properly the dmspectrum class
    2. dmflux          —> Show how-to compute gamma-ray flux for a specific target. Comparison with CLUMPY [Everything OK!]
3. pfiles/csdmatter.par
    1. Add decay as option for process
    2. Add logdfactor
    3. Add loglifetime
    4. Remove calc_t
    5. Comment fix_srcs and fix_bkg, as not needed
    6. Remove redshift and eblmodel input parameters
4. csdmatter.py
    1. stylish
    2. Add dictionary for available channels in PPPC4DMID files without EW corrections
    3. In __init__, self._masses is created as an empty list
    4. In __init__, add the class members jfactor, dfactor, sigmav and lifetime to retrieve reference parameters of the dark matter candidate. Initializatoin to zero before running the app where the I called the _get_parameters method.
    5. In query parameters: request parameters according to the dm process
    6. In _get_channel_key, recover number of channel according to dm process
    7. Add _gen_model to create GModelSky container and take care of the spectral part according to the dm-process
    8. In _fit_mass_point, define E_ref, E_min, and E_max according to the DM-process
    9. In _fit_mass_point, the calculation of theoretical and upper-limit integrated fluxes is computed up to 95% of the maximum available anergy (mass or mass/2 according to the case)
    10. In _fit_mass_point, show the GModels container after setting the DM model
    11. In _fit_mass_point, add e2flux and e2flux_err to dictionary with results for mass point
    12. In _fit_mass_point, add key to result according to dark-matter process: 
        1. DECAY
            1. lifetime_ref
            2. lifetime_lim
        2. ANNA
            1. sigmav_ref
            2. sigmav_lim
    13. In _fit_mass_point, recover catching of exceptions when running ctulimit instance
    14. In _fit_mass_point, calculation of exclusion limit for parameter of interest according to the process:
        1. DECAY:
            1. sigmav_UL = sigmav_ref * scale_factor
        2. ANNA:
            1. lifetime_UL = lifetime_ref / scale_factor
    15. In _fit_mass_point, When the user request a DiffuseSource, I was asking for GModelSpatialDiffuseCube, but the spatial component of the source used GModelSpatialDiffuseMap. I also update the mc_cone method to newer versions of gammalib
    16. In _fit_mass_point, recover of e2dPhi/dE and the associated error
    17. In _create_fits, add e2flux and e2flux_err columns
    18. In _create_fits, add paroi (parameter of interest) columns according to the dark-matter process

old/DMLimits.py
# ====================================================================
# This script perform the DM limits using the Cirelli et al. spectrum.
# ====================================================================

import ctools
import gammalib
import cscripts

import numpy as np

import os
import sys
import time
import argparse

from auxfunctions import *

#=====================#
# Routine entry point #
#=====================#

if __name__ == '__main__':

    options = argparse.ArgumentParser( description='This script compute \
        the DM limits using Cirelli et al. Spectrum' ,\
        fromfile_prefix_chars='@' )
    source = options.add_argument_group( 'Source' , \
        'All the relevant information about the source is passed in the inmodel\
        argument, where a XML file is passed describing parameters for the\
        observation, as position of the source, name, and spatial and spectral\
        models for the source. It is assumed that the file provides information about\
        the background model used in the observation. For more details, the user\
        is referred to the gammalib and ctools documentations :)' )
    source.add_argument( '--inmodel' ,\
        help='File with xml model' ,\
        type=str ,\
        required=True ,
        metavar='PerseusModel.xml' )
    source.add_argument( '--gname' ,\
        help='Name of the simulation' ,\
        type=str ,\
        required=True ,
        metavar='Perseus' )
    source.add_argument( '--ROIra' ,\
        help='Right Ascension of ROI (in degrees)'  ,\
        type=float ,\
        required=True ,\
        metavar='47.0' )
    source.add_argument( '--ROIdec' ,\
        help='Declination of ROI (in degrees)' ,\
        type=float ,\
        required=True ,\
        metavar='39.0' )
    source.add_argument( '--ROIradius' ,\
        help='Radius of ROI (in degrees)' ,\
        type=float ,\
        required=True ,\
        metavar='10.0' )

    instrument = options.add_argument_group( 'CTA-Intrument options' , \
        'Information about CTA IRF, observation time, etc.' )
    instrument.add_argument( '--pntra' ,\
        help='Right Ascension of pointing (in degrees)' ,\
        type=float ,\
        required=True ,\
        metavar='47.0' )
    instrument.add_argument( '--pntdec' ,\
        help='Declination of pointing (in degrees)' ,\
        type=float ,\
        required=True ,\
        metavar='39.0' )
    instrument.add_argument( '--pntradius' ,\
        help='Radius of the observation (in degrees)' ,\
        type=float ,\
        required=True ,\
        metavar='10.0' )
    instrument.add_argument( '--caldb' ,\
        help='Database production for the IRF file. \
            Options are: [ prod2 , prod3b-v1 , prod3b-v2 ] ' ,\
        type=str ,\
        default='prod3b-v2' ,\
        metavar='prod3b-v2' ,\
        choices=[ 'prod2' , 'prod3b-v1' , 'prod3b-v2' ] )
    instrument.add_argument( '--irf' ,\
        help='Instrument Response Function for CTA' ,\
        type=str ,\
        default='North_z20_5h' ,\
        metavar='North_z20_50h' )
    instrument.add_argument( '--hours' ,\
        help='Time for simulation of observation (in hours)' ,\
        type=float ,\
        default=5.0 ,\
        metavar='50.0' )
    instrument.add_argument( '--enumbins',\
        help='Number of energy bins' ,\
        type=int ,\
        default=10 ,\
        metavar='20' )
    instrument.add_argument( '--emin',\
        help='Minimum energy for events (TeV)' ,\
        type=float ,\
        default=0.03 ,\
        metavar='0.01' )
    instrument.add_argument( '--emax',\
        help='Maximum energy for events (TeV)' ,\
        type=float ,\
        default=10.0 ,\
        metavar='100.0' )
    instrument.add_argument( '--id' ,\
        help='A number identifier for events in simulation. ' ,\
        type=str ,\
        default='0' ,\
        metavar='0' )
    instrument.add_argument( '--pts' ,\
        help='Value of TS to compute Exclusion limits (TS < pts).' ,\
        type=float ,\
        default='25.0' ,\
        metavar='25.0' )
    instrument.add_argument( '--nsims' ,\
        help='Number of simulations' ,\
        type=int ,\
        required=True ,\
        metavar='10' )
    instrument.add_argument( '--nthreads' ,\
        help='Number of threads to speed the calculations' ,\
        type=int ,\
        metavar='10' ,\
        default=1 )
    instrument.add_argument( '--outpath' , \
        help='Path to save files' ,\
        type=str ,\
        default='./' ,\
        metavar='path/to/file/' )
    instrument.add_argument( '--is_onoff' ,\
        help='Indicate ON/OFF observation type' ,\
        action='store_true' )

    #   Create two parsers to manage the DM process selection
    subparsers = options.add_subparsers( help='Selecting DM processess \
        to produce gamma-rays' , dest='dmprocess' )
    subparsers.required = True

    #   Parser for annihilation
    anna       = subparsers.add_parser( 'anna' , help='DM annihilation Process' )
    anna.add_argument( '--sigmav' ,\
        help='Annihilation cross-section (cm**3/s)' ,\
        type=float ,\
        metavar='1.e-28' )
    anna.add_argument( '--jfactor' ,\
        help='Astrophysical J factor (in GeV**2/cm**5)' ,\
        type=float ,\
        metavar='1.e+20' )
    anna.add_argument( '--channel' ,\
        help='Annihilation channel (According to PPPC4DM Tables)' ,\
        type=int ,\
        metavar='11' ,\
        choices=[ 5 , 8 , 11 , 15 ] )
    anna.add_argument( '--mass' ,\
        help='DM mass (in TeV)' ,\
        type=float ,\
        metavar='100.0' )

    #   Parser for decay
    decs       = subparsers.add_parser( 'decs' , help='DM decay Process' )
    decs.add_argument( '--lifetime' ,\
        help='DM Decay Lifetime (s)' ,\
        type=float ,\
        metavar='1.e+35' )
    decs.add_argument( '--dfactor' ,\
        help='Astrophysical D factor (in GeV/cm**2)' ,\
        type=float ,\
        metavar='1.e+20' )
    decs.add_argument( '--channel' ,\
        help='Annihilation channel (According to PPPC4DM Tables)' ,\
        type=int ,\
        metavar='11' ,\
        choices=[ 5 , 8 , 11 , 15 ] )
    decs.add_argument( '--mass' ,\
        help='DM mass (in TeV)' ,\
        type=float ,\
        metavar='100.0' )


    #   Parsing arguments
    args = options.parse_args()

    MeVtoTeV = 1.e-6 ;

    #   This list put all the CTA-Background models
    bkgclasses = [ 'GCTAModelAeffBackground' , 'GCTAModelBackground' , \
        'GCTAModelCubeBackground' , 'GCTAModelIrfBackground' , \
        'GCTAModelRadialAcceptance' ]

    #   Check if xmlfile and outpath exist
    auxMan.checkDir( args.outpath )
    auxMan.checkPathandExit( args.inmodel )

    is_anna_or_decs = False
    process         = ''

    if args.dmprocess.lower() == 'anna' :
        is_anna_or_decs = True
        process         = 'Annihilation'

    elif args.dmprocess.lower() == 'decs' :

        is_anna_or_decs = False
        process         = 'Decay'

    #   Just printing useful information
    print( '************************************************' )
    print( '**    Mass:            {:.2e} TeV'.format( args.mass ) )
    print( '**    Process:         {:s}'.format( process ) )
    print( '**    Channel:         {:d}'.format( args.channel ) )

    if is_anna_or_decs :

        print( '**    Cross-Section:   {:.2e} cm**3/s'.format( args.sigmav ) )
        print( '**    J Factor:        {:.2e} GeV**2/cm**5'.format( args.jfactor ) )

    else :

        print( '**    Lifetime:        {:.2e} s'.format( args.lifetime ) )
        print( '**    D Factor:        {:.2e} GeV/cm**2'.format( args.dfactor ) )

    print( '************************************************' )

    #   Then, set a GModels container
    models = gammalib.GModels( args.inmodel )
    nameSources = [ source.name() \
        for source in models if source.classname() not in bkgclasses ]

    if args.gname in nameSources :

        print( '\n**    {:s} found in the models\n'.format( args.gname ) )

    else :

        print( '\n\tERROR: {:s} not found in Model container.\
            \n\tAvailbale models are:' )
        for name in nameSources :

            print( '\t - {:s}'.format( name ) )

        sys.exit()

    #   Extracting ts flag (calculation of TS)
    ts_flag = models[ args.gname ].tscalc()

    #   Extracting free parameters
    freepars = [ par.name() for par in models[ args.gname ].spectral() \
        if par.is_free() ]

    #   Set the GSkydir for pointing direction and source position
    pdir = gammalib.GSkyDir()
    rdir = gammalib.GSkyDir()
    pdir.radec_deg( args.pntra , args.pntdec )
    rdir.radec_deg( args.ROIra , args.ROIdec )

    #   Compute angular distance between directions
    dirs_sep = pdir.dist_deg( rdir )

    #   Creating two GSkyRegionCircle regions to check that
    #   Region of simulation contains ROI region
    pntregion = gammalib.GSkyRegionCircle( pdir , args.pntradius )
    roiregion = gammalib.GSkyRegionCircle( rdir , args.ROIradius )

    #   Check if pointing region contains ROI region
    is_inside = pntregion.contains( roiregion )

    if is_inside :

        print( ( "\n\nPonting region with:\n" +
               "\tRA     : {:.3f} deg\n".format( args.pntra ) +
               "\tDec    : {:.3f} deg\n".format( args.pntdec ) +
               "\tRadius : {:.2f} deg\n\n".format( args.pntradius ) +
               "contains ROI region.\n\nROI region:\n" +
               "\tRA     : {:.3f} deg\n".format( args.ROIra ) +
               "\tDec    : {:.3f} deg\n".format( args.ROIdec ) +
               "\tRadius : {:.2f} deg".format( args.ROIradius ) +
               "\n\nAngular separation between pointing and" +
               " position of the source is: {:.2f} deg\n".format( dirs_sep ) ) )

    else :

        print( ( "\n\nError:\n" +
               "Ponting region does not contain region associated with the source.\n" +
               "Please check the two directions.\n" +
               "\n\nAngular separation between pointing and"
               " position of the source is: {:.2f} deg\n".format( dirs_sep ) ) )

        sys.exit()


    #   Set duration of observation
    duration = args.hours * 3600.

    #   Set GEnergy objects and compute theoretical spectrum
    thiseref = 0.0

    if is_anna_or_decs :

        thiseref = args.mass / 2.0

    else :

        thiseref = args.mass / 4.0

    e_ref = gammalib.GEnergy( thiseref  , 'TeV' )
    e_min = gammalib.GEnergy( args.emin , 'TeV' )
    e_max = gammalib.GEnergy( args.emax , 'TeV' )

    #   Get expected dmflux at reference energy
    #   for all sources in Models
    theo_flux = 0.0

    for nameSource in nameSources :

        spec = models[ nameSource ].spectral()

        flux = spec.eval( e_ref , gammalib.GTime() )
        theo_flux += flux

    print( '\t* dF/dE(@ {:.2f} TeV): {:.3e} ph/(MeV cm**2 s)'.format( thiseref , theo_flux ) )

    #   Creating th fits file to save all the relevant results
    dmfits = gammalib.GFits()

    #   Creating table with number of rows equal to the number of sims
    table  = gammalib.GFitsBinTable( args.nsims )

    #   Creating columns for every parameter to save
    col_runID   = gammalib.GFitsTableShortCol( 'RunID' , args.nsims )
    col_events  = gammalib.GFitsTableDoubleCol( 'CubeEvents' , args.nsims )
    # col_oevents = gammalib.GFitsTableDoubleCol( 'ObsEvents' , args.nsims )
    col_ts      = gammalib.GFitsTableDoubleCol( 'TS' , args.nsims )
    col_ulflux  = gammalib.GFitsTableDoubleCol( 'ULFlux', args.nsims )
    col_scale   = gammalib.GFitsTableDoubleCol( 'ScaleFactor' , args.nsims )

    if is_anna_or_decs :

        col_cs  = gammalib.GFitsTableDoubleCol( 'LogCrossSection' , args.nsims )

    else :

        col_cs  = gammalib.GFitsTableDoubleCol( 'LogLifetime' , args.nsims )


    #   List of GFitsColumns
    colpars = []

    for freepar in freepars :

        col_par = gammalib.GFitsTableDoubleCol( freepar , args.nsims , 2 )
        colpars.append( col_par )

    #   At this moment I am not using this
    #   But, the idea is to try to use as an optional
    #   input parameter
    cubetype  = ''
    eventlist = ''


    #################################################################
    #####                       The Simuations
    #################################################################
    for nsim in range( args.nsims ) :

        #   ID to save in fits file
        thisrunID = args.id + '{:d}'.format( nsim )
        print( '\t**********************' )
        print( '\t*    Observation: {:d}'.format( nsim + 1 ) )
        print( '\t**********************' )

        #   GObservations Container
        obslist    = gammalib.GObservations()

        #   Setting observation parameters
        obs_binned = myCTAfuncs.single_obs( type=cubetype , eventfile=eventlist ,\
            pntdir=pdir , tstart=0.0 , duration=duration , deadc=0.95 ,\
            emin=args.emin , emax=args.emax , rad=args.pntradius ,\
            irf=args.irf , caldb=args.caldb , id=thisrunID , instrument='CTA' )

        #   Append obs_binned to obslist container
        obslist.append( obs_binned )

        #   And set the models to obslist container
        obslist.models( models )

        #   Name of the fits file to save events
        cubename = '{:s}_{:s}_{:d}_{:.2f}TeV_{:.2e}_Sim{:s}_{:d}min_cntcube.fits'.format( \
            args.gname , args.dmprocess , args.channel , args.mass ,\
            args.sigmav , thisrunID , int( duration ) )
        cubename = auxMan.createname( args.outpath , cubename )

        #   And now, perform the simulation using ctobssim
        #   and ctbin
        obssim = myCTAfuncs.sim( obslist , log=False , debug=False , \
            chatter=4 , edisp=False , seed=int( time.time() ) , \
            nbins=args.enumbins , binsz=0.02 , npix=200 , outfile=cubename )

        #   This is to save in the fits file
        events = obssim[ 0 ].events().number()

        #   Now, creating the observation definition XML file
        obsname = auxMan.createname( args.outpath , \
            'DMSimObs{:s}.xml'.format( thisrunID ) )

        obsxml  = gammalib.GXml()
        lists   = obsxml.append( 'observation_list title="observation library"' )
        obsss   = lists.append( 'observation name ="{:s}" id="{:s}"\
            instrument="CTA"'.format( args.gname , thisrunID ) )
        obsss.append( 'parameter name="EventList" file="{:s}"'.format( cubename ) )
        obsxml.save( obsname )

        #   Now, if the user request to generate an OnOff observation
        #   we use csphagen to handle this
        if args.is_onoff :

            print( "\t\tStarting calculation of On-Off files using csphagen app" )

            onoffname      = auxMan.createname( args.outpath , \
                '{:s}OnOffObs.xml'.format( thisrunID ) )
            onoffmodelname = auxMan.createname( args.outpath ,\
                '{:s}OnOffModel.xml'.format( thisrunID ) )
            prefix         = os.path.join( args.outpath ,\
                '{:s}OnOff'.format( thisrunID ) )

            onoffgen       = cscripts.csphagen()

            onoffgen[ 'inobs' ]      = obsname
            onoffgen[ 'inmodel' ]    = args.inmodel
            onoffgen[ 'srcname' ]    = args.gname
            onoffgen[ 'caldb' ]      = args.caldb
            onoffgen[ 'irf' ]        = args.irf
            onoffgen[ 'outobs' ]     = onoffname
            onoffgen[ 'outmodel' ]   = onoffmodelname
            onoffgen[ 'ebinalg' ]    = 'LOG'
            onoffgen[ 'emin' ]       = args.emin
            onoffgen[ 'emax' ]       = args.emax
            onoffgen[ 'enumbins' ]   = args.enumbins
            onoffgen[ 'coordsys' ]   = 'CEL'
            onoffgen[ 'ra' ]         = args.ROIra
            onoffgen[ 'dec' ]        = args.ROIdec
            onoffgen[ 'rad' ]        = args.ROIradius
            onoffgen[ 'bkgmethod' ]  = 'REFLECTED'
            onoffgen[ 'bkgregskip' ] = 1
            onoffgen[ 'bkgregmin' ]  = 2
            onoffgen[ 'etruemin' ]   = args.emin
            onoffgen[ 'etruemax' ]   = args.emax
            onoffgen[ 'nthreads' ]   = args.nthreads
            onoffgen[ 'stack' ]      = False
            onoffgen[ 'prefix' ]     = prefix

            onoffgen.execute()

            #   Number of observed events.
            nobserved = onoffgen.obs()[ 0 ].nobserved()

            print( ( "\n\t\tOnOff Model saved in:    {:s}".format( onoffmodelname ) +
                     "\n\t\tOnOff XML file saved in: {:s}".format( onoffname ) +
                     "\n\t\tMethod to create background regions: REFLECTED" ) )

        #   Now, Optimization using MLE
        print( '\t**    DM fitting' )

        likeModelName = auxMan.createname( args.outpath ,\
            '{:s}LikeOutModel.xml'.format( thisrunID ) )

        like = ctools.ctlike()

        if args.is_onoff :

            like[ 'inobs' ]     = onoffname
            like[ 'inmodel' ]   = onoffmodelname

        else :

            like[ 'inobs' ]     = obsname
            like[ 'inmodel' ]   = args.inmodel

        like[ 'caldb' ]         = args.caldb
        like[ 'irf' ]           = args.irf
        like[ 'outmodel' ]      = likeModelName
        like[ 'like_accuracy' ] = 1.e-4
        like[ 'max_iter' ]      = 100
        like[ 'nthreads' ]      = args.nthreads
        like[ 'debug' ]         = False

        print( '\t\t* Likelihood Calculation' )

        like.execute()

        #   Likelihood including DM Model
        #   (Alternative Hipotesis)
        like_alth = like.opt().value()

        #   Get the best fit value
        spec_fit  = like.obs().models()[ args.gname ].spectral()
        pars      = []

        for freepar in freepars :

            listpar = [ freepar , spec_fit[ freepar ].value() ,\
                spec_fit[ freepar ].error() ]
            pars.append( listpar )

        #Perform calculation of TS

        print( '\t\t* Getting value of TS' )

        TS = 0.0

        if not ts_flag :

            print( ( '\n\t\t* TS flag is not activated in XML model'
                     '\n\t\t* Calculation of TS using a copy of ctlike instance' ) )

            #   Now, removing the source model to
            #   compute the likelihood for Null hypothesis
            like.obs().models().remove( args.gname )
            like.run()

            #   Likelihood of null hypothesis
            like_null = like.opt().value()

            #   And, calculation of TS
            TS        = - 2.0 * ( like_alth - like_null )

        else :

            TS = like.obs().models()[ args.gname ].ts()

        print( '\n\t\t** TS = {:.3e}'.format( TS ) )

        #   Now, performing calculation of UL if TS < 25

        #  Initializing Scale factor and UL
        scale        = 0.0
        parameter_UL = 0.0
        diff_flux    = 0.0

        if TS < args.pts :

            print( '\t**    UL(95%C.L.) calculation' )

            limit = ctools.ctulimit()

            if args.is_onoff :

                limit[ 'inobs' ]   = onoffname
                limit[ 'inmodel' ] = onoffmodelname

            else :

                limit[ 'inobs' ]   = obsname
                limit[ 'inmodel' ] = args.inmodel

            limit[ 'srcname' ]     = args.gname
            limit[ 'eref' ]        = thiseref
            limit[ 'emin' ]        = args.emin
            limit[ 'emax' ]        = args.emax
            limit[ 'caldb' ]       = args.caldb
            limit[ 'irf' ]         = args.irf
            limit[ 'statistic' ]   = 'DEFAULT'

            limit.run()

            #   Get differential UL-flux
            diff_flux = limit.diff_ulimit()
            scale     = diff_flux / theo_flux

            print( ( '\n\t\t*** Upper limit of the differential flux:' +
                     ' {:.5e} cm**3/s'.format( diff_flux ) ) )

            if is_anna_or_decs :

                parameter_UL = args.sigmav * scale

                print( ( '\n\t\t*** Annihilation cross-section:' +
                         ' {:.5e} cm**3/s'.format( parameter_UL ) ) )

            else :

                parameter_UL = args.lifetime / scale

                print( '\n\t\t*** Decay Lifetime: {:.5e} s'.format( parameter_UL ) )

        #   Saving to fits file
        col_runID[ nsim ]  = nsim + 1
        col_events[ nsim ] = events
        col_ts[ nsim ]     = TS
        col_ulflux[ nsim ] = diff_flux

        #   If TS < 25, then save log(scale) and log(UL)
        #   else, save 0 for scale and UL
        if TS < args.pts :

            col_scale[ nsim ]  = np.log10( scale )
            col_cs[ nsim ]     = np.log10( parameter_UL )

        else :

            col_scale[ nsim ]  = 0.0
            col_cs[ nsim ]     = 0.0

        for ipar in range( len( pars ) ) :

            colpars[ ipar ][ nsim , 0 ] = pars[ ipar ][ 1 ]
            colpars[ ipar ][ nsim , 1 ] = pars[ ipar ][ 2 ]

        if nsim == 0 :

            print( 'Saving some files' )

        else :

            os.remove( obsname )
            os.remove( cubename )
            del obssim
            del obs_binned
            del obslist
            del obsxml

            if args.is_onoff :

                os.remove( onoffname )
                os.remove( onoffmodelname )

                os.remove( prefix + '_arf.fits' )
                os.remove( prefix + '_off.reg' )
                os.remove( prefix + '_on.reg' )
                os.remove( prefix + '_pha_off.fits' )
                os.remove( prefix + '_pha_on.fits' )
                os.remove( prefix + '_rmf.fits' )
                del onoffgen

            os.remove( likeModelName )
            del like

    print( '**    Finishing simulations' )

    #   Append columns to the table
    table.append( col_runID )
    table.append( col_events )
    table.append( col_ts )
    table.append( col_scale )
    table.append( col_cs )

    for ipar in range( len( colpars ) ) :

        table.append( colpars[ ipar ] )

    #   Cards to save information about the DM model
    table.card( 'OBJECT' , args.gname , 'Name of target' )

    #   Card for the DM process
    if is_anna_or_decs :

        table.card( 'DMPROCESS' , args.dmprocess , 'X(+) + X(-) -> gammas' )
        table.card( 'REF_CROSS_SECTION' , np.log10( args.sigmav ) ,\
            'log10(sigmav/[cm**3/s])' )

    else :

        table.card( 'DMPROCESS' , args.dmprocess , 'X -> gammas' )
        table.card( 'REF_LIFETIME' , np.log10( args.lifetime ) ,\
            'log10( lifetime/[s])' )

    table.card( 'CHANNEL' , args.channel , \
        'DM channel to generate gamma-rays' )
    table.card( 'DM_MASS' , args.mass , 'Mass of DM candidate (TeV)' )
    table.card( 'HOURS' , args.hours , 'Observation time (h)' )

    #   Then append to fits file
    dmfits.append( table )

    #   And save file
    fits_file = '{:s}_{:s}_{:s}_{:d}_Results.fits'.format( args.id , \
        args.gname ,args.dmprocess , args.channel )
    fits_file = auxMan.createname( args.outpath , fits_file )

    print( '** Results saved in: {:s}\n'.format( fits_file ) )

    dmfits.saveto( fits_file , True )

    print( 'This is the END!...' )
